### PR TITLE
faster content resolve

### PR DESCRIPTION
**Pull request by jrieken**
This is part of https://github.com/Microsoft/vscode/issues/37541 and reduces the waterfall by a few steps. It does that by reading the file only once and setting up the decoder-stream on the fly and by avoid a block stat-call whenever possible 

There are two big changes:

1. The new `resolveStringStream` method which by itself just stream contents. It handles encoding, binary files and bails out on very large file. Tho the latter should/can be managed from the caller. That is second part of this change...
2. which is in `resolveTextContent`. It builds an `IRawTextContent` object, ideally in parallel and only waterfall'ish when needed (etag compare). It also checks the size of the file and it will cancel a read-stream when a file is too large. Usually only a very small chunk of text has been loaded/wasted by that. I believe that the better comprise then also blocking on the stat-call.

This change makes the editor be the winner with normal sized files

nov-10-2017 17-41-24

### CODE REVIEW

File: src/vs/workbench/services/files/node/fileService.ts [hunk 1]
```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
 		return this.doResolveContent(resource, options, (stat, enc) => this.resolveFileContent(stat, enc));
 	}
 
+	private static chunkSize = 64 * 1024;
+	private static chunkBuffer = Buffer.allocUnsafe(FileService.chunkSize);
```
Comment (haacki47) [inline src/vs/workbench/services/files/node/fileService.ts]: Why did you do that as a `static`?
As i see you using this `field` within class, and method which takes value of `field` is not `static`.

`private chunkBuffer: type = Buffer.allocUnsafe(FileService.chunkSize);`

And don't forgot to assign type of `field`

File: src/vs/workbench/services/files/node/fileService.ts [hunk 2]
```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
...
+
+	public resolveStringStream(resource: uri, options?: IResolveContentOptions, token: CancellationToken = CancellationToken.None): IStringStream {
```
Comment (haacki47) [inline src/vs/workbench/services/files/node/fileService.ts]: This method i see like a **callback hell**
We need to do something with that.

File: src/vs/workbench/services/files/node/fileService.ts [hunk 3]
```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
...
+						ret.emit('error', err);
+					}
+					if (decoder) {
+						decoder.removeAllListeners();
+					}
+					ret.emit('end');
+				});
+			};
+
+			const handleChunk = (bytesRead) => {
```
Comment (haacki47) [inline src/vs/workbench/services/files/node/fileService.ts]: Put this logic to private separate method of this class.
Remove comments in conditions.

```typescript
private someAwesomeNameForMethod (bytesRead: type): void {
			if (token.isCancellationRequested) {
 					finish(new Error('cancelled'));
				} else if (bytesRead < FileService.chunkSize) {
					decoder.write(FileService.chunkBuffer.slice(0, bytesRead), finish);
				} else {
					decoder.write(FileService.chunkBuffer, readChunk);
				}
	}
```

Comment (haacki47) [reply src/vs/workbench/services/files/node/fileService.ts]: Then you can reuse this. And you don't need assign it to constant

File: src/vs/workbench/services/files/node/fileService.ts [hunk 4]
```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
...
+
+			if (err) {
+				ret.emit('error', err);
+				return;
+			}
+
+			let decoder: NodeJS.ReadWriteStream;
+			let totalBytesRead = 0;
+
+			const finish = (err?: any) => {
```
Comment (haacki47) [inline src/vs/workbench/services/files/node/fileService.ts]: Put this logic to separate method.

```typescript
private awesomeNameForMethod (err?: Error): void {
		if (err) {
					ret.emit('error', err);
			}
				fs.close(fd, err => {
					if (err) {
						ret.emit('error', err);
					} else if (decoder) {
						decoder.removeAllListeners();
					}
					ret.emit('end');
				});
	}
```

File: src/vs/workbench/services/files/node/fileService.ts [hunk 5]
```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
...
+					// done, write rest, end
+					decoder.write(FileService.chunkBuffer.slice(0, bytesRead), finish);
+
+				} else {
+					// read, write, repeat
+					decoder.write(FileService.chunkBuffer, readChunk);
+				}
+			};
+
+			const readChunk = () => {
```
Comment (haacki47) [inline src/vs/workbench/services/files/node/fileService.ts]: See comment above. Please keep your code clean.
And **DRY**

File: src/vs/workbench/services/files/node/fileService.ts [hunk 6]
```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
...
+							} else {
+								value.encoding = this.getEncoding(resource, this.getEncoding2(resource, options, value));
+								ret.emit('encoding', value.encoding);
+								decoder = encoding.decodeStream(value.encoding);
+								decoder.on('data', arg => ret.emit('data', arg));
+								decoder.on('error', arg => ret.emit('error', arg));
+								handleChunk(bytesRead);
+							}
+
+						}).catch(err => {
```
Comment (haacki47) [inline src/vs/workbench/services/files/node/fileService.ts]: Which type of `err`?

File: src/vs/workbench/services/files/node/fileService.ts [hunk 7]
```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
...
+						).then(value => {
+
+							if (options && options.acceptTextOnly && value.mimes.indexOf(baseMime.MIME_BINARY) >= 0) {
+								// Return error early if client only accepts text and this is not text
+								finish(new FileOperationError(
+									nls.localize('fileBinaryError', "File seems to be binary and cannot be opened as text"),
+									FileOperationResult.FILE_IS_BINARY
+								));
+
+							} else {
```
Comment (haacki47) [inline src/vs/workbench/services/files/node/fileService.ts]: A lots of instructions within `else`...
Make it more readable and simple.
Put some logic to methods.

File: src/vs/workbench/services/textfile/electron-browser/textFileService.ts [hunk 8]
```diff
@@ -55,20 +56,73 @@ export class TextFileService extends AbstractTextFileService {
...
-					mtime: streamContent.mtime,
-					etag: streamContent.etag,
-					encoding: streamContent.encoding,
-					value: res.value,
-					valueLogicalHash: res.hash
-				};
-				return r;
+
+		const result: IRawTextContent = {
+			resource: undefined,
```
Comment (haacki47) [inline src/vs/workbench/services/textfile/electron-browser/textFileService.ts]: What a type is `undefined`???
Are you sure that you need this value which will be always `undefined` in your program runtime

File: src/vs/workbench/services/textfile/electron-browser/textFileService.ts [hunk 9]
```diff
@@ -55,20 +56,73 @@ export class TextFileService extends AbstractTextFileService {
...
+			name: undefined,
+			mtime: undefined,
+			etag: undefined,
+			encoding: undefined,
+			value: undefined,
+			valueLogicalHash: undefined,
+		};
+
+		let contentToken = new CancellationTokenSource();
```
Comment (haacki47) [inline src/vs/workbench/services/textfile/electron-browser/textFileService.ts]: Specify type of this value

File: src/vs/workbench/services/files/node/fileService.ts [hunk 10]
```diff
@@ -623,6 +713,28 @@ export class FileService implements IFileService {
 		});
 	}
 
+	// TODO@ben, I extracted this from `resolveStreamContent` but I am not sure
+	// what it does in comparsion to the other encoding related functions...
+	public getEncoding2(resource: uri, options: IResolveContentOptions, detected: IMimeAndEncoding): string {
```
Comment (haacki47) [inline src/vs/workbench/services/files/node/fileService.ts]: We need to simplify this conditions. It looks like a hell

File: src/vs/workbench/services/textfile/electron-browser/textFileService.ts [hunk 11]
```diff
@@ -55,20 +56,73 @@ export class TextFileService extends AbstractTextFileService {
...
+			// a fresh load without a previous etag which means we can resolve the file stat
+			// and the content at the same time, avoiding the waterfall.
+			completePromise = Promise.all([statsPromise, this.fillInContents(result, resource, options, contentToken.token)]);
+		}
+
+		return TPromise.wrap(completePromise).then(() => result);
+	}
+
+	private fillInContents(content: IRawTextContent, resource: URI, options: IResolveContentOptions, token: CancellationToken): Thenable<any> {
+		const contentStream = this.fileService.resolveStringStream(resource, options, token);
```
Comment (haacki47) [inline src/vs/workbench/services/textfile/electron-browser/textFileService.ts]: Don't forgot specify type of value

File: src/vs/workbench/test/workbenchTestServices.ts [hunk 12]
```diff
@@ -738,6 +738,18 @@ export class TestFileService implements IFileService {
 		});
 	}
 
+	resolveStringStream(resource: URI, options?: IResolveContentOptions): IStringStream {
+		const callback: { [name: string]: Function } = Object.create(null);
```
Comment (haacki47) [inline src/vs/workbench/test/workbenchTestServices.ts]: For this case we can put `{[name: string]: Function}` to `custom-typings.d.ts`

```typescript
interface Dictionary<T extends any> {
    [key: string]: T;
}
```

Other developers will thank you for that.

File: src/vs/workbench/services/textfile/electron-browser/textFileService.ts [hunk 13]
```diff
@@ -55,20 +56,73 @@ export class TextFileService extends AbstractTextFileService {
...
+			etag: undefined,
+			encoding: undefined,
+			value: undefined,
+			valueLogicalHash: undefined,
+		};
+
+		let contentToken = new CancellationTokenSource();
+		let completePromise: Thenable<any>;
+
+		let statsPromise = this.fileService.resolveFile(resource).then(stat => {
```
Comment (bpasero) [inline src/vs/workbench/services/textfile/electron-browser/textFileService.ts]: I think this has to move into the `resolveStringStream` method because otherwise the `FILE_NOT_MODIFIED_SINCE` error will never be thrown for anyone that is using this method outside of the `textFileService`. If no `etag` is provided, the `stat` can run in parallel to the `fs.open` in `resolveStringStream` and otherwise we have to run it first.

The stat could then also be used within `resolveStringStream` to cancel reading the file in chunks if the file size is detected to be too large.

File: src/vs/workbench/services/textfile/electron-browser/textFileService.ts [hunk 14]
```diff
@@ -55,20 +56,73 @@ export class TextFileService extends AbstractTextFileService {
...
+		} else {
+			// a fresh load without a previous etag which means we can resolve the file stat
+			// and the content at the same time, avoiding the waterfall.
+			completePromise = Promise.all([statsPromise, this.fillInContents(result, resource, options, contentToken.token)]);
+		}
+
+		return TPromise.wrap(completePromise).then(() => result);
+	}
+
+	private fillInContents(content: IRawTextContent, resource: URI, options: IResolveContentOptions, token: CancellationToken): Thenable<any> {
```
Comment (bpasero) [inline src/vs/workbench/services/textfile/electron-browser/textFileService.ts]: What is the reason for not doing all of this from within the file service? We used to have a method `resolveStreamContent` that was used for the editor model loading that contains both stream and metadata and I think we can preserve all of that also with your solution. Imho the code here should be pushed down and we can just keep `resolveStreamContent` with your faster implementation?

File: src/vs/workbench/services/files/node/fileService.ts (hunk 10 above)
Comment (bpasero) [inline src/vs/workbench/services/files/node/fileService.ts]: This is just a way to prioritise all the different encodings that can be determined for a file (either by detecting the encoding, by settings or by user choice). There are some rules which encoding to pick in which case. I think the method can be `private`, not `public` and be renamed to something like `selectEncoding` or `pickEncoding` or `getPreferredEncoding`.

File: src/vs/workbench/services/files/electron-browser/fileService.ts [hunk 15]
```diff
@@ -168,6 +169,10 @@ export class FileService implements IFileService {
 		return this.raw.resolveContent(resource, options);
 	}
 
+	public resolveStringStream(resource: uri, options?: IResolveContentOptions, token?: CancellationToken): IStringStream {
```
Comment (bpasero) [inline src/vs/workbench/services/files/electron-browser/fileService.ts]: What does this new method mean for the remote file system implementation? Can a remote file system provide the same stream, I have not seen any updates of this for [`RemoteFileService`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/files/electron-browser/remoteFileService.ts#L79)

File: src/vs/workbench/services/files/node/fileService.ts [hunk 16]
```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
...
+
+			readChunk();
+		});
+
+		return ret;
+	}
+
+	//#endregion
+
 	public resolveStreamContent(resource: uri, options?: IResolveContentOptions): TPromise<IStreamContent> {
```
Comment (bpasero) [inline src/vs/workbench/services/files/node/fileService.ts]: This method is now not used anymore, I suggest to keep the interface and fold your work in `resolveStringStream` into this method.

File: src/vs/workbench/test/workbenchTestServices.ts [hunk 17]
```diff
@@ -738,6 +738,18 @@ export class TestFileService implements IFileService {
 		});
 	}
 
+	resolveStringStream(resource: URI, options?: IResolveContentOptions): IStringStream {
```
Comment (bpasero) [inline src/vs/workbench/test/workbenchTestServices.ts]: We need to ensure that the new code is covered by the same test suite that covers `resolveContent` today. There are many tests in [`fileService.test.ts`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/files/test/node/fileService.test.ts#L628) that cover various things (e.g. large file, not modified since, encoding).

**Review by bpasero — CHANGES_REQUESTED** [review]
@jrieken added some feedback to the code. the stream reading and handling seems correct to me.

Overall I feel that more of the code in `textFileService` should move into `fileService` because otherwise the method in `fileService` is not very useful for any other consumer.

File: src/vs/workbench/services/files/electron-browser/fileService.ts (hunk 15 above)
Comment (jrieken) [reply src/vs/workbench/services/files/electron-browser/fileService.ts]: In parts it actually already works like that. Tho, the 'old' stream function does do everything in once call. My plan is to nuke the old once this is done and then migrate the remote work. Makes sense?

File: src/vs/workbench/services/files/node/fileService.ts [hunk 18]
```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
 		return this.doResolveContent(resource, options, (stat, enc) => this.resolveFileContent(stat, enc));
 	}
 
+	private static chunkSize = 64 * 1024;
+	private static chunkBuffer = Buffer.allocUnsafe(FileService.chunkSize);
+
+	public resolveStringStream(resource: uri, options?: IResolveContentOptions, token: CancellationToken = CancellationToken.None): IStringStream {
+		const ret = new EventEmitter();
+
+		fs.open(resource.path, 'r', (err, fd) => {
```
[the remark this answers is not in the document]
Comment (jrieken) [reply src/vs/workbench/services/files/node/fileService.ts]: check

File: src/vs/workbench/services/files/node/fileService.ts [hunk 19]
```diff
@@ -195,6 +196,111 @@ export class FileService implements IFileService {
...
+
+			if (err) {
```
[the remark this answers is not in the document]
Comment (jrieken) [reply src/vs/workbench/services/files/node/fileService.ts]: check

File: src/vs/workbench/services/files/node/fileService.ts (hunk 10 above)
Comment (jrieken) [reply src/vs/workbench/services/files/node/fileService.ts]: check

File: src/vs/workbench/services/textfile/electron-browser/textFileService.ts (hunk 14 above)
Comment (jrieken) [reply src/vs/workbench/services/textfile/electron-browser/textFileService.ts]: > What is the reason for not doing all of this from within the file service?

No-one uses the file service like that, only the text file service... My understanding is that every other caller of `resolveContent` and `resolveContentStream` cares only about the actual data and the encoding. So, why make everyone pay for this? I'd prefer to see the file service as something more raw which the text file service uses to compose something more complex.

File: src/vs/workbench/services/files/electron-browser/fileService.ts (hunk 15 above)
Comment (bpasero) [reply src/vs/workbench/services/files/electron-browser/fileService.ts]: Yes.

File: src/vs/workbench/services/files/node/fileService.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (bpasero) [reply src/vs/workbench/services/files/node/fileService.ts]: @jrieken I just noticed that there is also the `FILE_NOT_FOUND` error that we return from [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/files/node/fileService.ts#L281). That should also be handled by checking for `ENOENT`.

File: src/vs/workbench/services/textfile/electron-browser/textFileService.ts (hunk 14 above)
Comment (bpasero) [reply src/vs/workbench/services/textfile/electron-browser/textFileService.ts]: I do not have a good argument other than a) making it harder to test this change if parts of it is spread between `fileService` and `textFileService` and b) making it harder to understand how the method behaves when called from the editor. I do think that something like the NOT_MODIFIED_SINCE is something that a remote file system couldn't implement as well (e.g. that was actually taken from the HTTP 304 Not Modified as inspiration).

**Comment by bpasero** [comment]
@jrieken after a longer review and some minor changes around error handling, I merged this in. I also added a test specifically for the case of a file larger 64kb, just in case. 

One thing I noticed is that we are not waiting for the `fs.close()` to finish when [finishing](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/files/node/fileService.ts#L375). I am not 100% sure about the consequences to be fair. When I look e.g. at a comparable method [`readExactlyByFile`](https://github.com/Microsoft/vscode/blob/master/src/vs/base/node/stream.ts#L58) the promise is wired up with the `fs.close` callback. 

Another thing I noticed is that in case of an error within the `resolveFileData` method, we never seem to cancel the token at all from within the method. I wonder if there could be a case where an error happens but we are still calling into `readChunk` or `handleChunk`?

And then last but not least, I now notice that dev tools "break on error" directly after window reload more often when a file cannot be found, I end up in the [`reject`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/files/node/fileService.ts#L347) case of the `fs.open` call. I guess that is a consequence of changing the code to be more promise based and dev tools insisting on breaking on rejected promises. 

image

**Comment by jrieken** [comment]
> One thing I noticed is that we are not waiting for the fs.close() to finish when finishing

It's a bit of a trick. Why wait for for close when all we want is content. The risk is that we won't see any error that happens during close... We can either make the operation wait for close, log the error to the console, or ignore it as today 🙈 

> Another thing I noticed is that in case of an error within the resolveFileData method, we never seem to cancel the token at all from within the method.

It should do the same via calling `finish`. So abort (not call `readChunk` again), end the stream, and propagate errors (iff any).

**Comment by bpasero** [comment]
@jrieken ok, I can add code to log the error at least and we see if it shows up after all. And yeah, looks like we never go into reading/chunking again after calling `finish`.

**Comment by bpasero** [comment]
@jrieken I am wondering if `fs.read()` would ever return `0` for `bytesread` even if the file is still not fully read and simply because no data was read (e.g. network drives?). I am seeing suspicious retry code in [`stream.readExactlyByFile` ](https://github.com/Microsoft/vscode/blob/master/src/vs/base/node/stream.ts#L89) which could be from a long time ago when maybe node.js behaved like that. 

Basically whenever we receive less bytes than we ask for, we stop reading. So far I could not figure out if this can ever happen, but maybe you came across this concern as well?

**Comment by jrieken** [comment]
Didn't encounter it but fair concern. The `ReadStream` itself only checks that it got some data: https://github.com/nodejs/node/blob/803cacd228c273214072a6b9ca98d8c83d217c97/lib/fs.js#L2102

**Comment by jrieken** [comment]
pushed https://github.com/Microsoft/vscode/commit/33acfb5182a76848946d261b2ef38b873fb864fe for that

**Comment by bpasero** [comment]
@jrieken to confirm the theory I debugged what this code does:

`fs.createReadStream('./test.txt').pipe(fs.createWriteStream('./copy.txt'));`

It looks like the stream ends as soon as `0` bytes are read, but there is never a check for the number of bytes read matching the size of the buffer, so your fix seems right to me.

For the case of `0` bytes read, we end up pushing a `null` buffer into the stream [here](https://github.com/nodejs/node/blob/master/lib/fs.js#L2115) and then this seems to trigger [`onEofChunk`](https://github.com/nodejs/node/blob/master/lib/_stream_readable.js#L220).
